File: src/main/rs/emulate/legacy/archive/ArchiveCodec.java
package rs.emulate.legacy.archive;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import rs.emulate.shared.util.CompressionUtils;
import rs.emulate.shared.util.DataBuffer;

/**
 * Contains methods for encoding and decoding {@link Archive}s.
 * 
 * @author Major
 * @author Graham
 */
public final class ArchiveCodec {

	/**
	 * Decodes the {@link Archive} in the specified {@link DataBuffer}.
	 * 
	 * @param buffer The buffer.
	 * @return The archive.
	 * @throws IOException If there is an error decompressing the data.
	 */
	public static Archive decode(DataBuffer buffer) throws IOException {
		if (buffer.isEmpty()) {
			return Archive.EMPTY_ARCHIVE;
		}

		int extractedSize = buffer.getUnsignedTriByte();
		int size = buffer.getUnsignedTriByte();
		boolean extracted = false;

		if (size != extractedSize) {
			buffer = CompressionUtils.bunzip2(buffer);
			extracted = true;
		}

		int count = buffer.getUnsignedShort();
		int[] identifiers = new int[count];
		int[] extractedSizes = new int[count];
		int[] sizes = new int[count];

		for (int entry = 0; entry < count; entry++) {
			identifiers[entry] = buffer.getInt();
			extractedSizes[entry] = buffer.getUnsignedTriByte();
			sizes[entry] = buffer.getUnsignedTriByte();
		}

		List<ArchiveEntry> entries = new ArrayList<>(count);

		for (int entry = 0; entry < count; entry++) {
			DataBuffer data = DataBuffer.allocate(extracted ? extractedSizes[entry] : sizes[entry]);
			data.fill(buffer);

			entries.add(new ArchiveEntry(identifiers[entry], extracted ? data : CompressionUtils.bunzip2(data)));
		}

		return new Archive(entries);
	}
	
	// TODO theres a bug with archive compression...

	/**
	 * Encodes the specified {@link Archive} into a {@link DataBuffer}.
	 * 
	 * @param archive The archive.
	 * @param compression The compression type (see {@link CompressionType}).
	 * @return The buffer.
	 * @throws IOException If there is an error compressing the archive.
	 */
	public static DataBuffer encode(Archive archive, CompressionType compression) throws IOException {
		try (ByteArrayOutputStream bos = new ByteArrayOutputStream(archive.getSize())) {
			List<ArchiveEntry> entries = archive.getEntries();
			int entryCount = entries.size();

			DataBuffer meta = DataBuffer.allocate(entryCount * (Integer.BYTES + 2 * 3) + Short.BYTES);
			meta.putShort(entryCount);

			for (ArchiveEntry entry : entries) {
				DataBuffer uncompressed = entry.getBuffer();
				DataBuffer compressed = CompressionUtils.bzip2(uncompressed);
				uncompressed.position(0); // We just read from this buffer, so reset the position.

				meta.putInt(entry.getIdentifier());
				meta.putTriByte(uncompressed.remaining());
				meta.putTriByte(compressed.remaining());

				switch (compression) {
					case ARCHIVE_COMPRESSION:
					case NO_COMPRESSION:
						bos.write(uncompressed.getRemainingBytes());
						break;
					case ENTRY_COMPRESSION:
						bos.write(compressed.getRemainingBytes());
						break;
					default:
						throw new IllegalArgumentException("Unrecognised compression type " + compression + ".");
				}
			}

			meta.flip();

			byte[] compressed = bos.toByteArray();
			DataBuffer data = DataBuffer.allocate(meta.limit() + compressed.length);
			data.put(meta).put(compressed).flip();

			DataBuffer header = DataBuffer.allocate(2 * 3);
			int extracted = data.limit();
			header.putTriByte(extracted);

			if (compression == CompressionType.ARCHIVE_COMPRESSION) {
				data = CompressionUtils.bzip2(data);
			}

			int compressedLength = data.limit();
			header.putTriByte(compressedLength).flip();

			DataBuffer buffer = DataBuffer.allocate(header.limit() + data.limit());
			buffer.put(header).put(data).flip();

			return buffer;
		}
	}

	/**
	 * Sole private constructor to prevent instantiation.
	 */
	private ArchiveCodec() {

	}

}

File: legacy/src/main/kotlin/rs/emulate/legacy/archive/ArchiveCodec.kt
package rs.emulate.legacy.archive

import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import rs.emulate.legacy.archive.CompressionType.ARCHIVE_COMPRESSION
import rs.emulate.legacy.archive.CompressionType.ENTRY_COMPRESSION
import rs.emulate.legacy.archive.CompressionType.NO_COMPRESSION
import rs.emulate.util.compression.bunzip2
import rs.emulate.util.compression.bzip2
import rs.emulate.util.readUnsignedTriByte
import rs.emulate.util.toByteArray
import rs.emulate.util.writeTriByte
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.util.ArrayList

/**
 * Contains methods for encoding and decoding [Archive]s.
 */
object ArchiveCodec {

    /**
     * Decodes the [Archive] in the specified [ByteBuf].
     * @throws IOException If there is an error decompressing the data.
     */
    fun decode(buffer: ByteBuf): Archive {
        var buffer = buffer
        if (buffer.readableBytes() == 0) {
            return Archive.EMPTY_ARCHIVE
        }

        val extractedSize = buffer.readUnsignedTriByte()
        val size = buffer.readUnsignedTriByte()
        var extracted = false

        if (size != extractedSize) {
            buffer = buffer.bunzip2(extractedSize)
            extracted = true
        }

        val count = buffer.readUnsignedShort()
        val identifiers = IntArray(count)
        val extractedSizes = IntArray(count)
        val compressedSizes = IntArray(count)

        for (entry in 0 until count) {
            identifiers[entry] = buffer.readInt()
            extractedSizes[entry] = buffer.readUnsignedTriByte()
            compressedSizes[entry] = buffer.readUnsignedTriByte()
        }

        val entries = ArrayList<ArchiveEntry>(count)

        for (entry in 0 until count) {
            val entrySize = if (extracted) extractedSizes[entry] else compressedSizes[entry]
            val data = buffer.readSlice(entrySize)
            val uncompressed = if (extracted) data else data.bunzip2(extractedSizes[entry])

            entries += ArchiveEntry(identifiers[entry], uncompressed)
        }

        return Archive(entries)
    }

    /**
     * Encodes the specified [Archive] into a [ByteBuf].
     * @throws IOException If there is an error compressing the archive.
     */
    fun encode(archive: Archive, compression: CompressionType): ByteBuf {
        ByteArrayOutputStream(archive.size).use { bos ->
            val entries = archive.entries
            val entryCount = entries.size

            val meta = Unpooled.buffer(entryCount * (Integer.BYTES + 2 * 3) + java.lang.Short.BYTES)
            meta.writeShort(entryCount)

            for (entry in entries) {
                val uncompressed = entry.buffer
                val compressed = uncompressed.bzip2()
                uncompressed.readerIndex(0) // We just read from this buffer, so reset the position

                meta.writeInt(entry.identifier)
                meta.writeTriByte(uncompressed.readableBytes())
                meta.writeTriByte(compressed.readableBytes())

                when (compression) {
                    ARCHIVE_COMPRESSION, NO_COMPRESSION -> bos.write(uncompressed.toByteArray())
                    ENTRY_COMPRESSION -> bos.write(compressed.toByteArray())
                }
            }

            val compressed = bos.toByteArray()
            var data = Unpooled.buffer(meta.readableBytes() + compressed.size)
            data.writeBytes(meta).writeBytes(compressed)

            val header = Unpooled.buffer(2 * 3)
            val extracted = data.readableBytes()
            header.writeTriByte(extracted)

            if (compression === ARCHIVE_COMPRESSION) {
                data = data.bzip2()
            }

            val compressedLength = data.readableBytes()
            header.writeTriByte(compressedLength)

            val buffer = Unpooled.buffer(header.readableBytes() + data.readableBytes())
            buffer.writeBytes(header).writeBytes(data)

            return buffer
        }
    }

}
